internal/nexodus/ctlconnectivty.go

```diff
@@ -0,0 +1,104 @@
+package nexodus
+
+import (
+	"encoding/json"
+	"fmt"
+	"go.uber.org/zap"
+	"net"
+)
+
+const batchSize = 10
+
+func (ac *NexdCtl) Connectivity(_ string, keepaliveResults *string) error {
+	res := ac.ax.connectivityProbe()
+	var err error
+
+	// Marshal the map into a JSON string.
+	keepaliveJson, err := json.Marshal(res)
+	if err != nil {
+		return fmt.Errorf("error marshalling connectivty results")
+	}
+
+	*keepaliveResults = string(keepaliveJson)
+
+	return nil
+}
+
+func (ax *Nexodus) connectivityProbe() map[string]KeepaliveStatus {
+	peerStatusMap := make(map[string]KeepaliveStatus)
+
+	if !ax.relay {
+		for _, value := range ax.deviceCache {
+			// skip the node sourcing the probe
+			if ax.wireguardPubKey == value.PublicKey {
+				continue
+			}
+			pubKey := value.PublicKey
+			nodeAddr := value.TunnelIp
+			// strip the /32 from the prefix if present
+			if net.ParseIP(value.TunnelIp) == nil {
+				nodeIP, _, err := net.ParseCIDR(value.TunnelIp)
+				nodeAddr = nodeIP.String()
+				if err != nil {
```

Comment:
This check should come right after `ParseCIDR()`. `nodeIP` can be `nil` here so you can't do `nodeIP.String()` unconditionally. I would just log an error and `continue`. In theory all of this info was validated before it went into the device cache so hopefully it never happens.

Reply:
Thanks! This was legacy when type checking was worse and the relay had a different value than the hosts. We can remove it completely.